### ISSUE TITLE
Consider secrets encryption at rest

### ISSUE THREAD

**Issue by iaguis** [kinvolk/lokomotive#662]
Kubernetes supports [encrypting secret data at rest](https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/).

However, simply enabling it is not very useful and only protects against etcd compromises, since the encryption key is stored in plain text on the disk so an attacker with host access can decrypt secrets.

To solve this we should use a [KMS provider](https://kubernetes.io/docs/tasks/administer-cluster/kms-provider/). On AWS there is an [encryption provider](https://github.com/kubernetes-sigs/aws-encryption-provider) that interfaces with the AWS KMS service. On bare-metal or packet we need another solution so we should investigate a bit more, maybe there's a way to use vault for this?

**Comment by invidian** [comment]
Don't you need to have etcd physically separated from kube-apiserver for this to make sense? So if someone takes over etcd server, they cannot read or modify Kubernetes data (outage still possible though).

**Comment by iaguis** [comment]
I don't think so. AFAIU the data is saved encrypted in etcd, so what an attacker with etcd access would read is encrypted data.

If we don't use a KMS provider an attacker with host access can get the key from disk and decrypt data in etcd. If we use a KMS provider the key is stored in the backend so an attacker can't decrypt data in etcd.

**Comment by invidian** [comment]
But KMS credentials will still be on the same host, so an attacker would have access to them anyway?

**Comment by iaguis** [comment]
On AWS, access to KMS can be can be tied to the node instance profile, and you can call the KMS methods to decrypt and encrypt. This allows protecting against to offline attacks, encryption keys to rotate frequently and auditing key access.

AFAIU with this, an attacker that gets access to the running host can still decrypt secrets by calling the KMS APIs but at least the keys are not stored plaintext on the host, access to KMS is audited, and if the node data is compromised when the node is offline attackers can't decrypt secrets.

This always makes me scratch my head a bit, there's a good explanation of [how this works in Google Cloud](https://cloud.google.com/kubernetes-engine/docs/how-to/encrypting-secrets).
